demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+}
+}
+
+CpuMonitor::CpuMonitor() : nCores{getNCores()}, lastEnabled{false}, historyEnabled{false}, samplesNumber{0} {};
+
+void CpuMonitor::openQuery() {
+    PDH_STATUS status;
+
+    status = PdhOpenQuery(NULL, NULL, &query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhOpenQuery() failed");
+    }
+
+    for (std::size_t i = 0; i < nCores; ++i)
+    {
+        TCHAR szCounterName[MAX_PATH];
+        _stprintf_s(szCounterName, sizeof(szCounterName), TEXT("\\Processor(%u)\\%% Processor Time"), i);
+
+        PDH_HCOUNTER counter;
+
+        status = PdhAddCounter(query, szCounterName, 0, &counter);
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhAddCounter() failed");
+        }
+
+        status = PdhSetCounterScaleFactor(counter, -2); // scale counter to [0, 1]
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhSetCounterScaleFactor() failed");
+        }
+
+        coreTimeCounters.push_back(counter);
+    }
+    status = PdhCollectQueryData(query);
+    if (ERROR_SUCCESS != status)
+    {
+        throw std::runtime_error("PdhCollectQueryData() failed");
+    }
+}
+
+void CpuMonitor::closeQuery() {
+    PDH_STATUS status = PdhCloseQuery(query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhCloseQuery() failed");
+    }
+}
+
+void CpuMonitor::enableHistory(std::size_t historySize) {
+    if (0 == historySize) {
+        disableHistory();
+    }
+    else {
+        this->historySize = historySize;
+        historyEnabled = true;
+        if (!lastEnabled) {
+            openQuery();
+        }
+    }
+}
+
+bool CpuMonitor::isHistoryEnabled() const {
+    return historyEnabled;
+}
+
+void CpuMonitor::disableHistory() {
+    historyEnabled = false;
+    if (!lastEnabled) {
+        closeQuery();
+    }
+}
+
+void CpuMonitor::enableLast() {
+    lastEnabled = true;
+    if (!historyEnabled) {
+        historySize = 1;
+        openQuery();
+    }
+}
+
+bool CpuMonitor::isLastEnabled() const {
+    return lastEnabled;
+}
+
+void CpuMonitor::disableLast() {
+    lastEnabled = false;
+    if (!historyEnabled) {
+        closeQuery();
+    }
+}
+
+void CpuMonitor::collectData() {
+    // store CPU Utulization in range [0, 1]
+    PDH_STATUS status;
+    PDH_FMT_COUNTERVALUE DisplayValue;
+
+    status = PdhCollectQueryData(query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhCollectQueryData() failed");
+    }
+
+    std::vector<std::pair<double, double>> idleNonIdleCpuStat;
+    for (std::size_t i = 0; i < coreTimeCounters.size(); i++) {
+        status = PdhGetFormattedCounterValue(coreTimeCounters[i], PDH_FMT_DOUBLE, NULL, &DisplayValue);
+        if (ERROR_SUCCESS != status) {
+            throw std::runtime_error("PdhGetFormattedCounterValue() failed");
+        }
+
+        idleNonIdleCpuStat.emplace_back(DisplayValue.doubleValue, 1 - DisplayValue.doubleValue);
```

Comment:
This doesn't quite make sense to me. You're filling `idleNonIdleCpuStat` with pairs, but then you only ever access the first item of those pairs. Why use pairs, then?

Also, you probably should check `DisplayValue.CStatus` before using `DisplayValue.doubleValue`.

Reply:
updated